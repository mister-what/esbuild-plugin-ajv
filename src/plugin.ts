import path from "path";
import fs from "fs/promises";
import { URL } from "url";
import type { Plugin } from "esbuild";
import type {
  CodeKeywordDefinition,
  Options as AjvOptions,
  AnySchemaObject,
} from "ajv";
import Ajv from "ajv";
import standaloneCode from "ajv/dist/standalone";
import { mergeDeepLeft } from "ramda";
import { createBuildCache } from "./cache";

export interface Options {
  extraKeywords?: CodeKeywordDefinition[];
  ajvOptions?: AjvOptions;
}

export const AjvPlugin = ({
  extraKeywords = [],
  ajvOptions = {},
}: Options = {}): Plugin => ({
  name: "ajv-plugin",
  setup(build) {
    const ajv = new Ajv(
      mergeDeepLeft(
        {
          code: {
            source: true,
            optimize: 1,
            esm: true,
            lines: true,
          },
          loadSchema: async (uri: string) =>
            JSON.parse(
              await fs.readFile(
                path.join(process.cwd(), new URL(uri).pathname),
                "utf8"
              )
            ) as AnySchemaObject,
        },
        ajvOptions
      )
    );

    for (const keywordDef of extraKeywords) {
      ajv.addKeyword(keywordDef);
    }

    const compileWithCache = createBuildCache(async (filePath, fileContent) => {
      ajv.removeSchema();
      const schema = JSON.parse(fileContent) as AnySchemaObject;
      schema.$id = `http://example.com/${path.relative(
        process.cwd(),
        filePath
      )}`;
      return standaloneCode(ajv, await ajv.compileAsync(schema));
    });

    build.onResolve(
      { filter: /\.json\?ajv$/i },
      async ({ path: rawPath, kind, resolveDir }) => {
        return {
          path: (
            await build.resolve(rawPath.replace(/\?ajv$/i, ""), {
              kind,
              resolveDir,
            })
          ).path,
          namespace: "ajv-validator",
        };
      }
    );
    build.onLoad(
      { namespace: "ajv-validator", filter: /.*/ },
      async ({ path: filePath }) => {
        return {
          contents: await compileWithCache(
            filePath,
            await fs.readFile(filePath, "utf8")
          ),
          loader: "js" as const,
          resolveDir: path.dirname(filePath),
        };
      }
    );
  },
});
