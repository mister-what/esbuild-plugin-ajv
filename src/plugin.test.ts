import path from "path";
import { runInThisContext } from "vm";
import * as esbuild from "esbuild";
import getTransformKeywordDef from "ajv-keywords/dist/definitions/transform";
import * as CodegenModule from "ajv/dist/standalone";
import type { Options } from "./plugin";
import { AjvPlugin } from "./plugin";
import * as BuildCacheModule from "./cache";

const wrapInCjs = (source: string) =>
  [
    // The second `function` wrapper is to prevent a strict mode error:
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Strict_Non_Simple_Params
    "(function({module, exports = module.exports, require, __dirname, __filename}) {(function(){",
    source,
    "})();});",
  ].join("\n");

const runBuild = async (
  outputFormat: "esm" | "cjs",
  options: esbuild.BuildOptions = {
    entryPoints: {
      main: path.resolve(__dirname, "__fixtures__", "test-entrypoint.js"),
    },
    outdir: path.resolve(__dirname, "build"),
  },
  ajvOptions: Options = {}
) => {
  const result = await esbuild.build({
    target: "node16",
    bundle: true,
    plugins: [
      AjvPlugin({
        ajvOptions: { coerceTypes: true },
        ...ajvOptions,
      }),
    ],
    format: outputFormat,
    minify: false,
    ...options,
    write: false,
  });

  return Object.fromEntries(
    result.outputFiles.map((out) => [
      path.relative(path.resolve(__dirname, "build"), out.path),
      Buffer.from(out.contents).toString("utf8"),
    ])
  )["main.js"];
};

const runBuildWithInput = (code: string, ajvOptions?: Options) =>
  runBuild(
    "cjs",
    {
      stdin: {
        contents: code,
        loader: "js",
        sourcefile: "test-entrypoint.js",
        resolveDir: path.resolve(__dirname, "__fixtures__"),
      },
      outfile: path.resolve(__dirname, "build", "main.js"),
    },
    ajvOptions
  );

const evaluateCjs = <M>(code: string): M => {
  const moduleObj = { exports: {} as M };
  const filename = "/test/file.js";
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  runInThisContext(wrapInCjs(code), { filename })({
    module: moduleObj,
    require: jest.fn(),
    __dirname: path.dirname(filename),
    __filename: filename,
  });
  return moduleObj.exports;
};

const makeIncrementalBuild = () => {
  const ajvPlugin = AjvPlugin({
    ajvOptions: { coerceTypes: true },
  });
  return esbuild.build({
    stdin: {
      contents: `
          import validator from "./testSchema.json?ajv";

          export const validate = (data) => {
            if (!validator(data)) return false;
            return data;
          };`,
      loader: "js",
      sourcefile: "test-entrypoint.js",
      resolveDir: path.resolve(__dirname, "__fixtures__"),
    },
    outfile: path.resolve(__dirname, "build", "main.js"),
    target: "node16",
    bundle: true,
    format: "esm",
    minify: false,
    write: false,
    plugins: [ajvPlugin],
    incremental: true,
  });
};

describe("AjvPlugin", () => {
  it("builds a a validation with external keywords", async () => {
    const { validate } = evaluateCjs<{ validate: (x: unknown) => unknown }>(
      await runBuildWithInput(
        // language=JavaScript
        `
          import validator from "./transformSchema.json?ajv";

          export const validate = (data) => {
            if (!validator(data)) return false;
            return data;
          };`,
        { extraKeywords: [getTransformKeywordDef()] }
      )
    );

    expect(
      validate({ text: "         padded text                  " })
    ).toEqual({ text: "padded text" });
    expect(validate({ text: 123 })).toEqual({ text: "123" });

    expect(validate({ text: "padded text", invalid: 123 })).toBe(false);
    expect(validate({ invalid: "text" })).toBe(false);
  });

  it("validates using the built schema", async () => {
    const { validate } = evaluateCjs<{ validate: (x: unknown) => unknown }>(
      await runBuild("cjs")
    );

    expect(() =>
      validate({ foo: "test", bar: 123, baz: [123, 456] })
    ).not.toThrow();

    expect(() =>
      validate({
        foo: "test",
        bar: 123,
        baz: [123, "abc", 789],
      })
    ).toThrow();
  });

  it("bundles built schemas", async () => {
    const source = await runBuild("esm");
    expect(source).toMatchCode({
      parser: "babel",
      trailingComma: "none",
      printWidth: 120,
    });
  });

  it("caches build results", async () => {
    const compileSpy = jest.spyOn(CodegenModule, "default");
    const createBuildCacheSpy = jest
      .spyOn(BuildCacheModule, "createBuildCache")
      .mockImplementation(
        (cb) =>
          (...args) =>
            Promise.resolve(cb(...args))
      );

    const result = await makeIncrementalBuild();
    expect(compileSpy).toHaveBeenCalledTimes(1);
    await result.rebuild?.();
    expect(compileSpy).toHaveBeenCalledTimes(2);
    result.rebuild?.dispose();

    createBuildCacheSpy.mockRestore();
    compileSpy.mockClear();

    const resultCached = await makeIncrementalBuild();
    expect(compileSpy).toHaveBeenCalledTimes(1);
    await resultCached.rebuild?.();
    expect(compileSpy).toHaveBeenCalledTimes(1);
    resultCached.rebuild?.dispose();
  });
});
