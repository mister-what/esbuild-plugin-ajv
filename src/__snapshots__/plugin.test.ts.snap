// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AjvPlugin bundles built schemas 1`] = `
var testSchema_default = validate11;
function validate11(data, { instancePath = "", parentData, parentDataProperty, rootData = data } = {}) {
  let vErrors = null;
  let errors = 0;
  if (errors === 0) {
    if (data && typeof data == "object" && !Array.isArray(data)) {
      let missing0;
      if ((data.foo === void 0 && (missing0 = "foo")) || (data.baz === void 0 && (missing0 = "baz"))) {
        validate11.errors = [
          {
            instancePath,
            schemaPath: "#/required",
            keyword: "required",
            params: { missingProperty: missing0 },
            message: "must have required property '" + missing0 + "'"
          }
        ];
        return false;
      } else {
        if (data.foo !== void 0) {
          let data0 = data.foo;
          const _errs1 = errors;
          if (typeof data0 !== "string") {
            let dataType0 = typeof data0;
            let coerced0 = void 0;
            if (!(coerced0 !== void 0)) {
              if (dataType0 == "number" || dataType0 == "boolean") {
                coerced0 = "" + data0;
              } else if (data0 === null) {
                coerced0 = "";
              } else {
                validate11.errors = [
                  {
                    instancePath: instancePath + "/foo",
                    schemaPath: "#/properties/foo/type",
                    keyword: "type",
                    params: { type: "string" },
                    message: "must be string"
                  }
                ];
                return false;
              }
            }
            if (coerced0 !== void 0) {
              data0 = coerced0;
              if (data !== void 0) {
                data["foo"] = coerced0;
              }
            }
          }
          var valid0 = _errs1 === errors;
        } else {
          var valid0 = true;
        }
        if (valid0) {
          if (data.bar !== void 0) {
            const _errs3 = errors;
            if (123 !== data.bar) {
              validate11.errors = [
                {
                  instancePath: instancePath + "/bar",
                  schemaPath: "#/properties/bar/const",
                  keyword: "const",
                  params: { allowedValue: 123 },
                  message: "must be equal to constant"
                }
              ];
              return false;
            }
            var valid0 = _errs3 === errors;
          } else {
            var valid0 = true;
          }
          if (valid0) {
            if (data.baz !== void 0) {
              let data2 = data.baz;
              const _errs4 = errors;
              if (errors === _errs4) {
                if (Array.isArray(data2)) {
                  var valid1 = true;
                  const len0 = data2.length;
                  for (let i0 = 0; i0 < len0; i0++) {
                    let data3 = data2[i0];
                    const _errs6 = errors;
                    if (!(typeof data3 == "number" && !(data3 % 1) && !isNaN(data3) && isFinite(data3))) {
                      let dataType1 = typeof data3;
                      let coerced1 = void 0;
                      if (!(coerced1 !== void 0)) {
                        if (
                          dataType1 === "boolean" ||
                          data3 === null ||
                          (dataType1 === "string" && data3 && data3 == +data3 && !(data3 % 1))
                        ) {
                          coerced1 = +data3;
                        } else {
                          validate11.errors = [
                            {
                              instancePath: instancePath + "/baz/" + i0,
                              schemaPath: "#/properties/baz/items/type",
                              keyword: "type",
                              params: { type: "integer" },
                              message: "must be integer"
                            }
                          ];
                          return false;
                        }
                      }
                      if (coerced1 !== void 0) {
                        data3 = coerced1;
                        if (data2 !== void 0) {
                          data2[i0] = coerced1;
                        }
                      }
                    }
                    var valid1 = _errs6 === errors;
                    if (!valid1) {
                      break;
                    }
                  }
                } else {
                  validate11.errors = [
                    {
                      instancePath: instancePath + "/baz",
                      schemaPath: "#/properties/baz/type",
                      keyword: "type",
                      params: { type: "array" },
                      message: "must be array"
                    }
                  ];
                  return false;
                }
              }
              var valid0 = _errs4 === errors;
            } else {
              var valid0 = true;
            }
            if (valid0) {
              if (data.external !== void 0) {
                let data4 = data.external;
                const _errs8 = errors;
                if (typeof data4 !== "string") {
                  let dataType2 = typeof data4;
                  let coerced2 = void 0;
                  if (!(coerced2 !== void 0)) {
                    if (dataType2 == "number" || dataType2 == "boolean") {
                      coerced2 = "" + data4;
                    } else if (data4 === null) {
                      coerced2 = "";
                    } else {
                      validate11.errors = [
                        {
                          instancePath: instancePath + "/external",
                          schemaPath: "./externalSchema.json#/definitions/name/type",
                          keyword: "type",
                          params: { type: "string" },
                          message: "must be string"
                        }
                      ];
                      return false;
                    }
                  }
                  if (coerced2 !== void 0) {
                    data4 = coerced2;
                    if (data !== void 0) {
                      data["external"] = coerced2;
                    }
                  }
                }
                var valid0 = _errs8 === errors;
              } else {
                var valid0 = true;
              }
            }
          }
        }
      }
    } else {
      validate11.errors = [
        { instancePath, schemaPath: "#/type", keyword: "type", params: { type: "object" }, message: "must be object" }
      ];
      return false;
    }
  }
  validate11.errors = vErrors;
  return errors === 0;
}
var validate = (value) => {
  if (!testSchema_default(value)) throw testSchema_default.errors;
  return value;
};
export { validate };

`;
